### ISSUE TITLE
Running value processor with Eclipse compiler silently fails to generate methods 

### ISSUE THREAD

**Issue by bhamiltoncx** [immutables/immutables#288]
Running Immutables.org `2.1.15` with the Eclipse compiler `ecj-4.5.1`, the value annotation processor silently fails to generate some methods in the generated classes.

I don't know what the smallest repro case is, but I've attached a repro case involving a Guava container of enums in a to-be-generated class.

Interestingly, the issue only reproduces in the Eclipse compiler, not in `javac 1.8.0_65`. I included commented-out code in the attached `build.sh` to show `javac` works fine.

To repro:

1) Unzip the attached `ecj-immutables.zip`
2) Run `./build.sh`

Expected results:

Compilation completes successfully both with `javac` and `ecj`.

Actual results:

Compilation completes successfully only with `javac`, and with `ecj` one of the methods fails to generate:

```
% ./build.sh                                                                     
+ '[' '!' -d gen ']'
+ mkdir gen
+ '[' '!' -d build ']'
+ java -jar jar/ecj-4.5.1.jar -1.7 -cp jar/value-2.1.15.jar:jar/guava-19.0.jar -processorpath jar/value-2.1.15.jar:jar/guava-19.0.jar -s gen -d build -sourcepath gen src/foo/Foo.java src/bar/Bar.java
----------
1. ERROR in /Users/bgertzfield/src/ecj-immutables/gen/foo/ImmutableFoo.java (at line 14)
	public final class ImmutableFoo extends Foo {
	                   ^^^^^^^^^^^^
The type ImmutableFoo must implement the inherited abstract method Foo.getFakeMultimap()
----------
1 problem (1 error)
```
[ecj-immutables.zip](https://github.com/immutables/immutables/files/182786/ecj-immutables.zip)

**Comment by bhamiltoncx** [comment]
I noticed we don't even emit the imports for `ImmutableList` when we use `ecj`:

```
% grep getFakeMultimap gen/foo/ImmutableFoo.java 
% grep ImmutableList gen/foo/ImmutableFoo.java  
% grep com.google gen/foo/ImmutableFoo.java   
import com.google.common.base.Preconditions;
```

**Comment by elucash** [comment]
Thanks again for well crafter repro cases!
The firsts results are that ECJ actually silently swallow those 2 methods, the one which returns the multimap and the one which takes immutable list of enum type. I mean annotation processing API is not receiving them at all. However as we have an access to the compiler internal structure we could trace that those declaration are present as AbstractMethodDeclarations, which however are not propagated to MethodBindings, being marked as broken/ignored. Overall it should be possible to reconstruct missing methods to some degree (i.e. with some level of detail) but it may be a significant undertaking.
So at this point, before diving into it, could we get away without it?

**Comment by bhamiltoncx** [comment]
Yikes. Well, we can start by filing a bug on the Eclipse team, of course

Is it possible I'm not invoking ecj correctly? I'm never quite sure what has to go on the classpath and what has to go on the processorpath.

How do people use Eclipse for annotation processing at all if something this fundamental is broken? I assume we can reproduce this in the Eclipse UI.

**Comment by elucash** [comment]
I'm afraid things like this is not specified precisely anywhere, and there are literally dozens of discrepancies between Javac and ECJ. You're invoking everything correctly. I'm still trying to see how it's possible to hijack this method resolution mechanism. There are may be following workarounds
1. Use `AbstractType.Enum` which is not nice aesthetically, but should work fine.
2. Use direct style of generation and extends yet to be generated builders by inner builder. Cons: somewhat increased boilerplate (still less than autovalue) and also in order to have withMethods without declaring corresponding abstract withers we need to introduce new feature of extending yet-to-be generated interface with `with*` methods declared. And also this will require to go over each value type to perform minor refinements.
3. Jump thru hoops and patch missing methods by combining source parsing with ECJ internal structures analysis.

**Comment by elucash** [comment]
Simple hijacking is not quite working (worked to recover non-abstract method but abstract one  `getFakeMultimap` still missing). I think it might be more beneficial to construct custom `ExecutableElement` implementation using the information recovered from ECJ's `MethodDeclaration`.
I'll will return to this is 1-2 days to try this approach.

**Comment by bhamiltoncx** [comment]
Any progress on this one?

**Comment by elucash** [comment]
Very little progress, sorry, got stuck in experimenting and was overwhelmingly busy. Will return to this tomorrow to see how feasible is to implement elements from scratch using declaration objects mentioned above.

**Comment by elucash** [comment]
Just as an update: trying to figure out the strategy how to (re-)construct missing elements, there are 2 options here, either to stub elements from scratch (the complication here is that it is hard to construct annotation mirrors), or to use Eclipse compiler internals to construct partially. I started with the first options and stuck at annotation handling. But I hope there should be a way out.

**Comment by elucash** [comment]
@bhamiltoncx an update on this, or rather some thoughts. Each time I'm getting to this and trying different approaches I end up with way to many custom code and hacks and still not getting complete resolution. I'm thinking of giving up on this mainly because the workaround would be quite easy: use `AbstractType.Enum` with generic declarations like `Map<AbstractType.Enum, String>` instead of `Map<Type.Enum, String>`.

Otherwise this bug is draining too much of energy. 2.2.beta is going to be released in a couple of days with generics support, nullable collections and default attributes etc. I'm also starting to work on a next gen version of Immutables for java 8 (while maintaining 2.x for foreseeable future). This new version would be: lessons learned + some innovation + realize the vision for gadt and optics. So to speak in this new version it would be natural to use `Type.Enum` where `Type` is both abstract value type and a canonical value type, so the kind of solution to this bug would be just unnecessary.

**Comment by bhamiltoncx** [comment]
I totally understand it's a really tricky bug. Maybe we could come up with
a repro case to file on the Eclipse team?

I'd definitely rather you spend your time working on the next version.
2016年4月28日(木) 23:13 Eugene <notifications@github.com>:

**Comment by elucash** [comment]
Here's the bug report https://bugs.eclipse.org/bugs/show_bug.cgi?id=492761 Not sure if it will get any traction, but, I guess, we did our best.

**Comment by elucash** [comment]
Just tested with the latest ECJ v4.6M7, the bug is fixed there.

**Comment by bhamiltoncx** [comment]
That's fantastic! I'll test it out on our side.

**Comment by bhamiltoncx** [comment]
Yep, bug is fixed with ECJ v4.6M7. Can't get their jar to work with ant just yet but I'll figure that out separately. Thanks for following through!
